uaclient/files/user_config_file.py

```diff
@@ -0,0 +1,134 @@
+        Field("poll_for_pro_license", BoolDataValue, required=False),
+        Field("polling_error_retry_delay", IntDataValue, required=False),
+        Field("metering_timer", IntDataValue, required=False),
+        Field("update_messaging_timer", IntDataValue, required=False),
+    ]
+
+    def __init__(
+        self,
+        apt_http_proxy: Optional[str] = None,
+        apt_https_proxy: Optional[str] = None,
+        global_apt_http_proxy: Optional[str] = None,
+        global_apt_https_proxy: Optional[str] = None,
+        ua_apt_http_proxy: Optional[str] = None,
+        ua_apt_https_proxy: Optional[str] = None,
+        http_proxy: Optional[str] = None,
+        https_proxy: Optional[str] = None,
+        apt_news: Optional[bool] = None,
+        apt_news_url: Optional[str] = None,
+        poll_for_pro_license: Optional[bool] = None,
+        polling_error_retry_delay: Optional[int] = None,
+        metering_timer: Optional[int] = None,
+        update_messaging_timer: Optional[int] = None,
+    ):
+        self.apt_http_proxy = apt_http_proxy
+        self.apt_https_proxy = apt_https_proxy
+        self.global_apt_http_proxy = global_apt_http_proxy
+        self.global_apt_https_proxy = global_apt_https_proxy
+        self.ua_apt_http_proxy = ua_apt_http_proxy
+        self.ua_apt_https_proxy = ua_apt_https_proxy
+        self.http_proxy = http_proxy
+        self.https_proxy = https_proxy
+        self.apt_news = apt_news
+        self.apt_news_url = apt_news_url
+        self.poll_for_pro_license = poll_for_pro_license
+        self.polling_error_retry_delay = polling_error_retry_delay
+        self.metering_timer = metering_timer
+        self.update_messaging_timer = update_messaging_timer
+
+
+event = event_logger.get_event_logger()
+LOG = logging.getLogger(util.replace_top_level_logger_name(__name__))
+
+
+class UserConfigFileObject:
+    def __init__(self, directory: str = defaults.DEFAULT_DATA_DIR):
+        file_name = defaults.USER_CONFIG_FILE
+        self._private = DataObjectFile(
+            UserConfigData,
+            UAFile(
+                file_name,
+                os.path.join(directory, defaults.PRIVATE_SUBDIR),
+                private=True,
+            ),
+            DataObjectFileFormat.JSON,
+            optional_type_errors_become_null=True,
+        )
+        self._public = DataObjectFile(
+            UserConfigData,
+            UAFile(file_name, directory, private=False),
+            DataObjectFileFormat.JSON,
+            optional_type_errors_become_null=True,
+        )
+
+    def redact_config_data(
+        self, user_config: UserConfigData
+    ) -> UserConfigData:
+        redacted_data = copy.deepcopy(user_config)
+        for field in PROXY_FIELDS:
+            value = getattr(redacted_data, field)
+            if value:
+                setattr(
+                    redacted_data,
+                    field,
+                    "<REDACTED>",
+                )
```

Comment:
I think we only need to redact if there is a username or password present. `urlparse` can be used to check for for the existence of those fields, then we can just redact the whole URL if one of them is present. What do you think @lucasmoura @dheyay?

Reply:
should be good I think. 

do we have cases where root user set's the proxy value and it does not contain username, password? In that case this should be fine as long as we want the non-root user to be able to see such proxy values. @orndorffgrant

Reply:
Yes that is exactly the scenario I think we should support